python/hsfs/engine/spark.py

```diff
@@ -405,14 +405,17 @@ def write_training_dataset(
             else:
                 raise ValueError("Dataset should be a query.")
 
+            if training_dataset.coalesce:
+                dataset = dataset.coalesce(1)
+
+            dataset = dataset.cache()
```

Comment:
Why cache here?

Reply:
I think I did this to cache the query. If we do the split we cache the split result instead. I would have liked to cache the query as well when we split, but I was a bit concerned about the potential memory consumption when caching before and after the split.

Reply:
If there is no split, I think we should not cache the result and users can cache themselves. The purpose of caching is to return consistent result.

Reply:
Well okay, the reasoning behind it was to prevent 2x execution in the case of 1. having transformation fuctions and need to calcualte statistics and 2. have to write the df to disk. But I agree that placing the caching here without checking for this particular case is not great. Shall we still keep it for the special case though?

Reply:
Ah right. Need to calculate statistics. We should cache then.

Reply:
I created a Jira here to add it later https://hopsworks.atlassian.net/browse/FSTORE-317